### PR TITLE
Extend WasmHttpMessageHandler

### PR DESCRIPTION
**Pull request by Suchiman**
- Adds support for HttpCompletionOption.ResponseHeadersRead by delaying reading of the response body into HttpContent
- Adds optional support for streaming responses in supported browsers
- Adds support for canceling in flight fetch requests

### CODE REVIEW

**Comment by kjpou1** [comment]
[The following two lines](https://github.com/mono/mono/pull/11255/files#diff-ac5d67d0aeb34ebf13b83109de2c32b6L172):

```
                httpresponse = null;
                
                status.Dispose();
```

Should not be removed.  This will introduce a memory leak.  To see the memory leak one can request an http request in a loop and view afterward the memory.

What happens is that all the internal plumbing that is allocated in the HttpRequest; for example Headers, references to streams, arrays, and text, etc..; will never be garbage collected if it is not set to null.  As well as `status`.  When GC kicks in the associated memory will be released.


```
       static int requests = 0;

        private static async void HttpClientRequest()
        {

            string ApiFile = "SampleJPGImage_30mbmb.jpg";

            try
                {
                    var rspMsg = await httpClient.GetAsync(ApiFile);
                        Console.WriteLine($"Request: {++requests}  StreamAsync: {rspMsg.Content?.ReadAsStreamAsync().Result.Length}");
                }
                catch (Exception exc2)
                {
                    Console.WriteLine($"Playground : {exc2.Message}");
                    Console.WriteLine($"Playground inner: {exc2.InnerException?.Message}");
                }

            Console.WriteLine($"Total Memory End HttpClientRequest: Request: {requests} WA: {getTotalMemory()}, GC: {GC.GetTotalMemory(false)}");

        }

```
example output:

```
mono.js:2347 Request: 308  StreamAsync: 30789588
mono.js:2347 Total Memory End HttpClientRequest: Request: 308 WA: 268435456, GC: 65939176
mono.js:2347 GC_MAJOR_SWEEP: major size: 864K in use: 120K
mono.js:2347 GC_MAJOR: (LOS overflow) time 45.30ms, stw 45.30ms los size: 0K in use: 0K
mono.js:2347 streaming supported: True
mono.js:2347 streaming enabled: False
mono.js:2347 Request: 309  StreamAsync: 30789588
mono.js:2347 Total Memory End HttpClientRequest: Request: 309 WA: 268435456, GC: 65939352
mono.js:2347 GC_MAJOR_SWEEP: major size: 864K in use: 121K
mono.js:2347 GC_MAJOR: (LOS overflow) time 44.60ms, stw 44.70ms los size: 0K in use: 0K

```

Without those two lines a memory error is generated after around 30 requests on my system :

```

Request: 30  StreamAsync: 30789588
mono.js:2347 Total Memory End HttpClientRequest: Request: 30 WA: 2133196800, GC: 1851711728
mono.js:11657 Cannot enlarge memory arrays. Either (1) compile with  -s TOTAL_MEMORY=X  with X higher than the current value 2133196800, (2) compile with  -s ALLOW_MEMORY_GROWTH=1  which allows increasing the size at runtime, or (3) if you want malloc to return NULL (0) instead of this abort, compile with  -s ABORTING_MALLOC=0 
mono.js:11658 Cannot enlarge memory arrays. Either (1) compile with  -s TOTAL_MEMORY=X  with X higher than the current value 2133196800, (2) compile with  -s ALLOW_MEMORY_GROWTH=1  which allows increasing the size at runtime, or (3) if you want malloc to return NULL (0) instead of this abort, compile with  -s ABORTING_MALLOC=0 

```

I suspect the same thing happens in Blazor in the [BrowserHttpMessageHandler](https://github.com/aspnet/Blazor/blob/master/src/Microsoft.AspNetCore.Blazor.Browser/Http/BrowserHttpMessageHandler.cs#L89).

**Comment by Suchiman** [comment]
The disposal of `_status` was moved into the `WasmHttpContent ` / ` WasmHttpReadStream` classes. If you dispose the `rspMsg`, this will also `Dispose` the `Content` and that will `Dispose`, `_status`.
Regarding `httpresponse = null;`, i don't know about the mono GC, but in .NET you don't need to set locals that go out of scope to null to have them be eligible for garbage collection.

File: sdks/wasm/WasmHttpMessageHandler.cs [hunk 1]
```diff
@@ -166,34 +187,29 @@ private async Task doFetch(TaskCompletionSource<HttpResponseMessage> tcs, HttpRe
                     ));
                 }
 
-
                 tcs.SetResult(httpresponse);
-
-                httpresponse = null;
```
[the remark this answers is not in the document]
Comment (kjpou1) [reply sdks/wasm/WasmHttpMessageHandler.cs]: Looks like multiple messages got posted: https://blog.github.com/2018-10-21-october21-incident-report/

[the remark this answers is not in the document]
Comment (kjpou1) [reply sdks/wasm/WasmHttpMessageHandler.cs]: Looks like multiple messages got posted: https://blog.github.com/2018-10-21-october21-incident-report/

[the remark this answers is not in the document]
Comment (kjpou1) [reply sdks/wasm/WasmHttpMessageHandler.cs]: https://github.com/mono/mono/pull/11255/files

**Comment by kjpou1** [comment]
[The following two lines](https://github.com/mono/mono/pull/11255/files#diff-ac5d67d0aeb34ebf13b83109de2c32b6L172):

```
                httpresponse = null;
                
                status.Dispose();
```

Should not be removed.  This will introduce a memory leak.  To see the memory leak one can request an http request in a loop and view afterward the memory.

What happens is that all the internal plumbing that is allocated in the HttpRequest; for example Headers, references to streams, arrays, and text, etc..; will never be garbage collected if it is not set to null.  As well as `status`.  When GC kicks in the associated memory will be released.


```
       static int requests = 0;

        private static async void HttpClientRequest()
        {

            string ApiFile = "SampleJPGImage_30mbmb.jpg";

            try
                {
                    var rspMsg = await httpClient.GetAsync(ApiFile);
                        Console.WriteLine($"Request: {++requests}  StreamAsync: {rspMsg.Content?.ReadAsStreamAsync().Result.Length}");
                }
                catch (Exception exc2)
                {
                    Console.WriteLine($"Playground : {exc2.Message}");
                    Console.WriteLine($"Playground inner: {exc2.InnerException?.Message}");
                }

            Console.WriteLine($"Total Memory End HttpClientRequest: Request: {requests} WA: {getTotalMemory()}, GC: {GC.GetTotalMemory(false)}");

        }

```
example output:

```
mono.js:2347 Request: 308  StreamAsync: 30789588
mono.js:2347 Total Memory End HttpClientRequest: Request: 308 WA: 268435456, GC: 65939176
mono.js:2347 GC_MAJOR_SWEEP: major size: 864K in use: 120K
mono.js:2347 GC_MAJOR: (LOS overflow) time 45.30ms, stw 45.30ms los size: 0K in use: 0K
mono.js:2347 streaming supported: True
mono.js:2347 streaming enabled: False
mono.js:2347 Request: 309  StreamAsync: 30789588
mono.js:2347 Total Memory End HttpClientRequest: Request: 309 WA: 268435456, GC: 65939352
mono.js:2347 GC_MAJOR_SWEEP: major size: 864K in use: 121K
mono.js:2347 GC_MAJOR: (LOS overflow) time 44.60ms, stw 44.70ms los size: 0K in use: 0K

```

Without those two lines a memory error is generated after around 30 requests on my system :

```

Request: 30  StreamAsync: 30789588
mono.js:2347 Total Memory End HttpClientRequest: Request: 30 WA: 2133196800, GC: 1851711728
mono.js:11657 Cannot enlarge memory arrays. Either (1) compile with  -s TOTAL_MEMORY=X  with X higher than the current value 2133196800, (2) compile with  -s ALLOW_MEMORY_GROWTH=1  which allows increasing the size at runtime, or (3) if you want malloc to return NULL (0) instead of this abort, compile with  -s ABORTING_MALLOC=0 
mono.js:11658 Cannot enlarge memory arrays. Either (1) compile with  -s TOTAL_MEMORY=X  with X higher than the current value 2133196800, (2) compile with  -s ALLOW_MEMORY_GROWTH=1  which allows increasing the size at runtime, or (3) if you want malloc to return NULL (0) instead of this abort, compile with  -s ABORTING_MALLOC=0 

```

I suspect the same thing happens in Blazor in the [BrowserHttpMessageHandler](https://github.com/aspnet/Blazor/blob/master/src/Microsoft.AspNetCore.Blazor.Browser/Http/BrowserHttpMessageHandler.cs#L89).

**Comment by Suchiman** [comment]
The fix to prevent this from happening would be making `JSObject` finalizable.
`JSObject` already calls `GC.SuppressFinalize(this);` in it's `Dispose` method but it does not implement a Finalizer.

**Comment by kjpou1** [comment]
I now see the `status.Dispose()` and that is fine because it is moved into `WasmHttpContent`/`WasmHttpReadStream` classes.

The `httpresponse = null` is because of the use of the `httpresponse` in the code:

```

                    respHeaders.Invoke("forEach", new Action<object, object, object>((value, name, other) =>
                    {

```

from @kumpera 

> if a local is captured (used) by a lambda it becomes heap memory as we translate them into fields on an object.

So actually my initial comment about Blazor does not hold up because it does not use a lambda when loading the headers as I initially thought.

**Comment by kjpou1** [comment]
Hello Robin @suchiman

There is a problem with canceling inflight streaming requests.  When the cancellation token is canceled the `AbortController` has already been `Disposed()`.  The culprit is coming from `WasmHttpReadStream.ReadAsync` where the _status is being disposed  https://github.com/mono/mono/pull/11255/files#diff-ac5d67d0aeb34ebf13b83109de2c32b6R364

**Comment by kjpou1** [comment]
@monojenkins build

**Comment by Suchiman** [comment]
A few things to note here about `Fix cancellation`
When no `Timeout` is set and no custom `CancellationToken` is given, a per `HttpClient` `CancellationTokenSource` is being used (which triggers when either disposing `HttpClient` or calling `CancelPendingRequests`) which means we have to make sure to `Dispose` / unregister our delegates we `Register` or else we're going to leak them.

When using `HttpCompletionOption.ResponseHeadersRead` and we are not using the per `HttpClient` `CancellationToken`, a LinkedTokenSource is passed to `SendAsync` which gets disposed immediately after `HttpResponseMessage` has been returned. Without `HttpCompletionOption.ResponseHeadersRead`, `HttpClient` will buffer the response and the `CancellationToken` will stay active for this duration. That's why i didn't run into this case in my tests.

Lastly i moved the disposal of `status` to when we've reached the end of the `Stream` to avoid running into the cancellation issues.

**Comment by kjpou1** [comment]
Robin

Thanks for looking at this and the explanations.  Will give this a look over.

**Comment by kjpou1** [comment]
This is excellent Robin @Suchiman !!!

Thank you for the contribution.

**Comment by kjpou1** [comment]
@monojenkins build

**Pull request merged**